File: internal/types/manifest.go
// internal/types/manifest.go
package types

import "time"

// Manifest represents archived data metadata for a feed/date
type Manifest struct {
	Feed             string       `json:"feed" yaml:"feed"`
	Date             string       `json:"date" yaml:"date"`
	Format           string       `json:"format" yaml:"format"`
	RotationInterval string       `json:"rotation_interval" yaml:"rotation_interval"`
	Files            []FileEntry  `json:"files" yaml:"files"`
	Gaps             []Gap        `json:"gaps" yaml:"gaps"`
	Tickers          []string     `json:"tickers" yaml:"tickers"`
	MessageTypes     []string     `json:"message_types" yaml:"message_types"`
	HasGaps          bool         `json:"has_gaps" yaml:"has_gaps"`
}

// FileEntry represents a single archived file
type FileEntry struct {
	Name         string    `json:"name" yaml:"name"`
	Start        time.Time `json:"start" yaml:"start"`
	End          time.Time `json:"end" yaml:"end"`
	Records      uint64    `json:"records" yaml:"records"`
	Bytes        uint64    `json:"bytes" yaml:"bytes"`
	NatsStartSeq uint64    `json:"nats_start_seq" yaml:"nats_start_seq"`
	NatsEndSeq   uint64    `json:"nats_end_seq" yaml:"nats_end_seq"`
}

// Gap represents a detected gap in the data stream
type Gap struct {
	AfterSeq     uint64    `json:"after_seq" yaml:"after_seq"`
	MissingCount uint64    `json:"missing_count" yaml:"missing_count"`
	DetectedAt   time.Time `json:"detected_at" yaml:"detected_at"`
}

// TotalRecords returns the sum of records across all files
func (m *Manifest) TotalRecords() uint64 {
	var total uint64
	for _, f := range m.Files {
		total += f.Records
	}
	return total
}

// TotalBytes returns the sum of bytes across all files
func (m *Manifest) TotalBytes() uint64 {
	var total uint64
	for _, f := range m.Files {
		total += f.Bytes
	}
	return total
}

File: ssmd-rust/crates/ssmd-archiver/src/manifest.rs
use chrono::{DateTime, Utc};
use serde::{Deserialize, Serialize};

#[derive(Debug, Serialize, Deserialize)]
pub struct Manifest {
    pub feed: String,
    pub date: String,
    pub format: String,
    pub rotation_interval: String,
    pub files: Vec<FileEntry>,
    pub gaps: Vec<Gap>,
    pub tickers: Vec<String>,
    pub message_types: Vec<String>,
    pub has_gaps: bool,
}

#[derive(Debug, Serialize, Deserialize, Clone)]
pub struct FileEntry {
    pub name: String,
    pub start: DateTime<Utc>,
    pub end: DateTime<Utc>,
    pub records: u64,
    pub bytes: u64,
    #[serde(skip_serializing_if = "Option::is_none", default)]
    pub raw_bytes: Option<u64>,
    #[serde(skip_serializing_if = "Option::is_none", default)]
    pub compression_ratio: Option<f64>,
    pub nats_start_seq: u64,
    pub nats_end_seq: u64,
}

#[derive(Debug, Serialize, Deserialize, Clone)]
pub struct Gap {
    pub after_seq: u64,
    pub missing_count: u64,
    pub detected_at: DateTime<Utc>,
}

impl Manifest {
    pub fn new(feed: &str, date: &str, rotation_interval: &str, format: &str) -> Self {
        Self {
            feed: feed.to_string(),
            date: date.to_string(),
            format: format.to_string(),
            rotation_interval: rotation_interval.to_string(),
            files: Vec::new(),
            gaps: Vec::new(),
            tickers: Vec::new(),
            message_types: Vec::new(),
            has_gaps: false,
        }
    }
}
